pubtools/tracing.py

```diff
@@ -0,0 +1,123 @@
+"""Instrument Pub task functions.
+
+Usage:
+    @instrument_func()
+      def func():
+          pass
+
+"""
+import functools
+import logging
+import os
+import threading
+
+from opentelemetry import baggage, context, trace
+from opentelemetry.baggage.propagation import W3CBaggagePropagator
+from opentelemetry.exporter.otlp.proto.http.trace_exporter import \
+    OTLPSpanExporter
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+from opentelemetry.trace import Status, StatusCode
+from opentelemetry.trace.propagation.tracecontext import \
+    TraceContextTextMapPropagator
+
+propagator = TraceContextTextMapPropagator()
+baggage_propagator = W3CBaggagePropagator()
+log = logging.getLogger(__name__)
+
+
+class TracingWrapper:
+    """Wrapper class that will wrap all methods of calls with the instrument_tracing decorator."""
+
+    __instance = None
+
+    def __new__(cls):
+        """Create a new instance if one does not exist."""
+        if not os.getenv("PUB_OTEL_TRACING", "").lower() == "true":
+            return None
+
+        if TracingWrapper.__instance is None:
+            log.info("Creating TracingWrapper instance")
+            cls.__instance = super().__new__(cls)
+            otlp_exporter = OTLPSpanExporter(
```

Comment:
I don't think that the exporter part of this should be hardcoded to use only this type of exporter.

Correct me if I'm wrong, but I think the intended design is that you can set up all your tracing code, and then choose different exporters for different scenarios without having to change the tracing code?

e.g. in https://opentelemetry.io/docs/instrumentation/python/exporters/#console-exporter it refers to using a console exporter which seems to be built in to the main otel library.  It says it's good for "development and debugging tasks" and I can imagine that indeed being quite useful for us.

For instance, if I'm working on pubtools-foo and I'm adding instrumentation code, it would be really nice if I can somehow make it use the console exporter when I run commands locally, but then use the HTTP exporter when deployed in Pub.  I wouldn't want to be forced to deploy some kind of OTLP/HTTP receiver locally just to test my OTEL-related changes during local development.

I think you should consider using a hook to create the exporter.  You can look at existing get_cert_key_paths hook to get an idea what's needed across pubtools & pub for that.

So, it could be something like this:

```python
@hookspec(firstresult=True)
def otel_exporter():
    """Return an OTEL exporter, used by OTEL instrumentation.

    If OTEL tracing is enabled and this hook is not implemented, a default
    `ConsoleSpanExporter` will be used.
    """
```

...then in Pub, implement the otel_exporter hook to return an OTLPSpanExporter.

That should allow all the OTEL related code to work even in local development and tests while exporting to console, and work in Pub by using OTLP/HTTP.

Reply:
Added this hook.